### PR TITLE
[learn] introduction/active-record revise and update

### LINKED ISSUE #230: Rewrite "ActiveRecord and ROM"
It should be now feasible to rewrite this section and turn it into a side-by-side comparison.

### PR DESCRIPTION
**Pull request by abrthel**
Fixes #230

I'm basically done here other than some wording changes and a few more examples. As always feel free to change anything and/or burn it all to the ground 🔥 😄 whatever gets a better result.

I had a lot of trouble not being "combative" towards ActiveRecord so definitely keep that in mind when reviewing and making changes.

### CODE REVIEW

File: source/4.0/learn/introduction/active-record.html.md [hunk 1]
```diff
@@ -1,82 +1,545 @@
...
+
+<h4 class="text-center">Active Record</h4>
+```ruby
+User.where("admin IS ? OR moderator IS ?", true, true)
+```
+
+
+<h4 class="text-center">ROM</h4>
+```ruby
+users_relation.where { admin.is(true).or(moderator.is(true)) }
```
Comment (flash-gordon) [inline source/4.0/learn/introduction/active-record.html.md]: there's also `|`, `where { admin.is(true) | moderator.is(true) }`

**Comment by abrthel** [comment]
Thanks for the feedback @flash-gordon! I'll look over these tonight and update 😄

File: source/4.0/learn/introduction/active-record.html.md [hunk 2]
```diff
@@ -1,82 +1,545 @@
...
+  4. [Persistence](#persistence)
+  6. [Validation](#validation)
+  7. [Where ROM Shines](#where-rom-shines)
+$TOC
 
-## Database Support
+Rails ActiveRecord is the most popular persistence
+framework in ruby land. Deployed inside the majority of rails applications
+across the web, it provides API's for quick and simple data access making it a
+great good solution for constructing CRUD style applications.
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: I suspect you wanted to write either "great" or "good", not both :)

File: source/4.0/learn/introduction/active-record.html.md [hunk 3]
```diff
@@ -1,82 +1,545 @@
...
 
-#### Active Record
+Our intention for this guide is to act as a primer for anyone familiar with
+Rails' ActiveRecord and looking for a quick start guide. Examples in each set will
+show how ActiveRecord accomplishes each task followed by an example
+with the equivalent using ROM.
 
-Supports only SQL.
+All ROM examples are based on `rom-sql` which is an adapter needed to use
+SQL databases with ROM. Information on installing on installing and
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: redundant "on installing"

File: source/4.0/learn/introduction/active-record.html.md [hunk 4]
```diff
@@ -1,82 +1,545 @@
...
-Data access logic is part of the object and controls all reading and writing to
-the database. You use the same objects to create, read, update, and delete data.
-These objects are the models in traditional Rails applications.
+## Models vs Relations
 
-#### ROM
+The first difference is ROM doesn't really have a concept of models. ROM objects
+are instantiated by the mappers and have no knowledge about persistence. You can
+map to whatever structure you want and in common use-cases you can use
+repositories to automatically map query results to simple struct-like objects.
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: This should be `...use **relations** to automatically...`

File: source/4.0/learn/introduction/active-record.html.md [hunk 5]
```diff
@@ -1,82 +1,545 @@
...
 
-Data manipulation is handled by a separate interface with user defined commands.
-Every relation that your application is going to use is explicitly defined. The
-ROM relations expose powerful internal query APIs that you use to create
-publicly accessible relation methods to return query results.
+The closest implementation to models would be would be `ROM::Struct`, which is
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: redundant "would be"

File: source/4.0/learn/introduction/active-record.html.md [hunk 6]
```diff
@@ -1,82 +1,545 @@
...
+essentially a nice hash with callable attributes and the ability to add custom
+behavior. More on ROM Structs later.
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: I'd say "essentially a data object with attribute readers, coercible to a hash"

File: source/4.0/learn/introduction/active-record.html.md (hunk 1 above)
Comment (solnic) [reply source/4.0/learn/introduction/active-record.html.md]: actually, there IS NO `or` (yet) :/

File: source/4.0/learn/introduction/active-record.html.md [hunk 7]
```diff
@@ -1,82 +1,545 @@
...
+```
+
+It should be noted that updating a record in ActiveRecord generally requires 
+that record to first be loaded then updated then committed. We view this as
+a bad practice as it leads to more round trips from the database and entities
+that are initialized in an invalid state. If a developer is sufficiently
+validating data at the boundaries of the application then updating or creating
+a record without loading it should be no problem and in fact preferable.
+
+### Create Nested Objects
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: Let's drop this section for now. We can add it once rom 5.0 is done, where nested changesets will work too.

File: source/4.0/learn/introduction/active-record.html.md [hunk 8]
```diff
@@ -1,82 +1,545 @@
...
+end
 
-There's no validation concept built-in. Validations are handled externally by
-separate libraries and validated data can be passed down to the command layer to
-be persisted.
+users.map_with(:encryption)
+```
 
-## Data Coercion
+### Transform Data Before Persisting
```
Comment (solnic) [inline source/4.0/learn/introduction/active-record.html.md]: This section could use a bit of introduction. Changesets with data mapping is a first class feature, and personally I find it to be REALLY DAMN USEFUL, and it's something AR does not have at all :)

**Comment by solnic** [comment]
This is fantastic work man. I left some comments. If you could address them, I'll follow up with a couple additions too. Then we could merge it in, assuming we're all happy with the final result. Again, outstanding work, thank you! <3

File: source/4.0/learn/introduction/active-record.html.md [hunk 9]
```diff
@@ -1,82 +1,545 @@
...
+  For information on how to configure a ROM environment see either
+  [Setup DSL](/%{version}/learn/getting-started/setup-sql)
+  or
+  [Rails Setup](/%{version}/learn/getting-started/rails-setup)
+  guides.
+^
 
-## Data Access
+^
+  Both frameworks have many similar API's but philosophically they are
```
[the remark this answers is not in the document]
Comment (abrthel) [reply source/4.0/learn/introduction/active-record.html.md]: `APIs` is correct. I have a bad habit of throwing a ' before my 's'-es. I especially screw it up when writing it's vs its

File: source/4.0/learn/introduction/active-record.html.md (hunk 1 above)
Comment (abrthel) [reply source/4.0/learn/introduction/active-record.html.md]: Huh.. I swear I tested that code sample. Ah well, must be crazy

File: source/4.0/learn/introduction/active-record.html.md (hunk 4 above)
Comment (abrthel) [reply source/4.0/learn/introduction/active-record.html.md]: Whoops, not sure what I was thinking there. Must have been drunk 😄

File: source/4.0/learn/introduction/active-record.html.md (hunk 7 above)
Comment (abrthel) [reply source/4.0/learn/introduction/active-record.html.md]: I've just commented out this section but left it in place for when it gets updated.

File: source/4.0/learn/introduction/active-record.html.md (hunk 8 above)
Comment (abrthel) [reply source/4.0/learn/introduction/active-record.html.md]: I've written up a little diddy for the beginning here but if you have some examples to flesh out the section a bit more that  would be super helpful.

**Comment by abrthel** [comment]
Thanks Solnic & Flash for the feedback, if I've learned anything while writing these docs is that I need an editor to review my writing before it ever sees the light of day.

As for the PR, suggested changes have been applied.

**Pull request merged**